File: c#/exceptions/ExceptionsHierarchy.cs
class NumberValueException : System.Exception {
  public int Number;
  public new string Message = "";

  public NumberValueException(int number) {
    Number = number;
    Message = "Some number has beign given.";
  }
}

class ZeroException : NumberValueException {
  public ZeroException(int number) : base(number) {
    Message = "0 number has beign given.";
  }
}

class OneException : NumberValueException {
  public OneException(int number) : base(number) {
    Message = "1 number has beign given.";
  }
}

class ThousandException : NumberValueException {
  public ThousandException(int number) : base(number) {
    Message = "1000 number has beign given.";
  }
}

class ExceptionsHierarchy {
  static int drawNumber() {
    System.Console.Write("Give some number: ");
    int number = System.Convert.ToInt32(System.Console.ReadLine());

    if (number == 0) {
      throw new ZeroException(number);
    } else if (number == 1) {
      throw new OneException(number);
    } else if (number == 10) {
      throw new NumberValueException(number);
    } else if (number == 1000) {
      throw new ThousandException(number);
    }

    return number;
  }

  static void Main(string[] args) {
    System.Console.WriteLine("Program begin...");

    try {
      System.Console.WriteLine("Risky code...");

      int number = drawNumber();

      System.Console.WriteLine("Given number " + number + " didn't case exception throwing.");
    } catch (ZeroException e) {
      System.Console.WriteLine("CASE 1: " + e.Message + " (" + e.Number + ")");
    } catch (OneException e) {
      System.Console.WriteLine("CASE 2: " + e.Message + " (" + e.Number + ")");
    } catch (NumberValueException e) {
      System.Console.WriteLine("CASE 0: " + e.Message + " (" + e.Number + ")");
    } //catch (ThousandException e) {
      // System.Console.WriteLine("CASE 3: " + e.Message + " (" + e.Number + ")");
    // }

    System.Console.WriteLine("Program end...");
  }
}

File: python/exceptions/exceptions_hierarchy.py
#!/usr/bin/python3

class NumberValueException(BaseException):
    number = 0
    message = ""

    def __init__(self, number):
        self.number = number
        self.message = "Some number " + str(self.number) + " has beign given."

class ZeroException(NumberValueException):
    def __init__(self, number):
        self.number = number
        self.message = "0 number has beign given."

class OneException(NumberValueException):
    def __init__(self, number):
        self.number = number
        self.message = "1 number has beign given."

class ThousandException(NumberValueException):
    def __init__(self, number):
        self.number = number
        self.message = "1000 number has beign given."


def draw_number():
    number = int(input("Give some number: "))

    if number == 0:
        raise ZeroException(number)
    elif number == 1:
        raise OneException(number)
    elif number == 10:
        raise NumberValueException(number)
    elif number == 1000:
        raise ThousandException(number)

    return number

print("Program begin...")

try:
    print("Risky code...")

    number = draw_number()

    print("Given number", number, "didn't case exception throwing.")
except ZeroException as e:
    print("CASE 1: " + e.message + " (" + str(e.number) + ")")
except OneException as e:
    print("CASE 2: " + e.message + " (" + str(e.number) + ")")
except NumberValueException as e:
    print("CASE 0: " + e.message + " (" + str(e.number) + ")")
except ThousandException as e: # Will be never catched here.
    print("CASE 3: " + e.message + " (" + str(e.number) + ")")

print("Program end...")
